rfcs/97-add-restrictions-in-storage-metadata.md

```diff
@@ -0,0 +1,148 @@
+
+- `copy-size-maximum`: Maximum size for copy operation.
+- `fetch-size-maximum`: Maximum size for fetch operation.
+- `move-size-maximum`: Maximum size for move operation.
+- `write-size-maximum`: Maximum size for write operation.
+- `append-number-maximum`: Max append numbers in append operation.
+- `append-size-maximum`: Max append size in per append operation.
+- `append-total-size-maximum`: Max append total size in append operation.
+- `multipart-number-maximum`: Maximum part number in multipart operation.
+- `multipart-size-maximum`: Maximum part size defined by storage service.
+- `multipart-size-minimum`: Minimum part size defined by storage service.
+
+### Deprecate the Existing Storage Metadata in Object Metadata
+
+The following append and multipart related metadata in the current object metadata, actually belongs to storage metadata, should be deprecated in object metadata.
+
+```go
+type Object struct {
+	...
+    // AppendNumberMaximum Max append numbers in append operation
+    appendNumberMaximum int
+    // AppendSizeMaximum Max append size in per append operation
+    appendSizeMaximum int64
+    // AppendTotalSizeMaximum Max append total size in append operation
+    appendTotalSizeMaximum int64
+	...
+    // MultipartNumberMaximum Maximum part number in multipart operation
+    multipartNumberMaximum int
+    // MultipartSizeMaximum Maximum part size defined by storager
+    multipartSizeMaximum int64
+    // MultipartSizeMinimum Minimum part size defined by storager
+    multipartSizeMinimum int64
+	...
+}
+```
+
+
+Based on above, the added storage metadata and corresponding `set/get` functions could be generated and added into `StorageMeta`.
+
+```go
+type StorageMeta struct {
+	location               string
+	Name                   string
+	WorkDir                string
+	copySizeMaximum        int64
+	fetchSizeMaximum       int64
+	moveSizeMaximum        int64
+	writeSizeMaximum       int64
+	appendNumberMaximum    int64
+	appendSizeMaximum      int64
+	appendTotalSizeMaximum int64
+	multipartNumberMaximum int64
+	multipartSizeMaximum   int64
+	multipartSizeMinimum   int64
+	
+	// bit used as a bitmap for object value, 0 means not set, 1 means set 
+	bit uint64
+	m   map[string]interface{}
+}
+```
+
+For services:
+- A member with type `StorageMeta` SHOULD be added, and the separate fields should be removed in `Storage`, like the following.
+- The storage metadata `sm` should be initiated at the time the storager initiated.
+
+```go
+type Storage struct {
+	sm           StorageMeta
+	
+	defaultPairs DefaultStoragePairs
+	features     StorageFeatures
+	
+	typ.UnimplementedStorager
+	...
+}
+```
+
+## Rationale
+
+### Alternative Way: Ignore size limit
+
+The alternative way is to send request directly without size check and get an error when the `size` larger than the limit.
+
+This will need invalid request and get the ambiguous error.
+
+## Compatibility
+
+This change will break all services.
```

Comment:
Can we provide a way for migration?

After mark `Object.appendNumberMaximum` as deprecated, I think this change will not break services?

Reply:
Yes, this change will not break services.
Maybe services need to bump the major version as `Storage` strcut will be modified?

Reply:
Not all changes should bump major versions.

Changes that don't need bump:

- Mark as deprecated
- Add new fields

Changes that need bump:

- Delete fields
- Rename fields

To me, those fields are not widely used (only in dm), so we just need to make sure the `go build` is not break.

Maybe we can migrate in the following ways:

- Add fields in `StorageMeta` and mark `Object.appendNumberMaximum` as deprecated
- Release a new version for `go-storage` and requires all service bump to this version with all references to `Object.appendNumberMaximum` updated
- Remove deprecated fields in `Object` (After coordinate with the whole community or next major version)

Reply:
Got it.
For the servives, is it neccessary to add a member with type `StorageMeta` in `Storage`? Maybe we just need to return the added fields in `metadata`?

Reply:
Both OK to me, your decision.

Reply:
I discard changes on `Storage` as it will add complexity to use storage metadata internally.